src/App.tsx

```diff
@@ -1,24 +1,235 @@
+import { TodoRow } from './components/TodoRow';
+import { TodoHeader } from './components/TodoHeader';
+import { Status } from './types/Status';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorMessage } from './components/ErrorMessage';
+
+const getFilteredTodos = (todos: Todo[], isCompleted: Status): Todo[] => {
+  return todos.filter((todo: Todo) => {
+    switch (isCompleted) {
+      case Status.Completed:
+        return todo.completed;
+      case Status.Active:
+        return !todo.completed;
+      default:
+        return true;
+    }
+  });
+};
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [status, setStatus] = useState<Status>(Status.All);
+
+  useEffect(() => {
+    todoService
+      .getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const timerId = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerId.current) {
+      window.clearTimeout(timerId.current);
+    }
+
+    timerId.current = window.setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  }, [errorMessage]);
+
+  const filterTodosIds = (newId: number) => {
+    setProcessingTodoIds(prevTodoIds => prevTodoIds
+      .filter(id => id !== newId));
+  };
+
+  const setNewIds = (newId: number) => {
+    setProcessingTodoIds(prevTodoIds => [...prevTodoIds, newId]);
+  };
+
+  const handleAddTodo = (todoTitle: string) => {
+    setTempTodo({
+      id: 0,
+      title: todoTitle,
+      userId: 0,
+      completed: false,
+    });
+
+    return todoService
+      .addTodos(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setNewIds(todoId);
+
+    todoService
+      .deleteTodos(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        filterTodosIds(todoId);
+      });
+  };
+
+  const handleRenameTodo = (todo: Todo, newTodoTitle: string) => {
+    setNewIds(todo.id);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        title: newTodoTitle,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => {
+          return currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo;
+        }));
```

Comment:
This logic is the same in the function below, maybe better to create function for this?